pupil_src/shared_modules/marker_tracker_3d/optimization/model_optimizer.py

```diff
@@ -0,0 +1,75 @@
+import background_helper
+from marker_tracker_3d import utils
+from marker_tracker_3d.model_optimizer_storage import ModelOptimizerStorage
+from marker_tracker_3d.optimization.optimization_generator import optimization_generator
+from marker_tracker_3d.optimization.visibility_graphs import VisibilityGraphs
+
+
+class ModelOptimizer:
+    def __init__(self, camera_model, update_menu=None):
+        self.camera_model = camera_model
+
+        self.model_optimizer_storage = ModelOptimizerStorage()
+        self.origin_marker_id = None
+        self.visibility_graphs = VisibilityGraphs(
+            self.model_optimizer_storage,
+            self.camera_model,
+            self.origin_marker_id,
+            update_menu,
+        )
+
+        self.bg_task = None
+
+    def update(self, marker_detections, camera_extrinsics):
+        self.visibility_graphs.add_marker_detections(
+            marker_detections, camera_extrinsics
+        )
+
+        self._run_optimization()
+
+        return self._get_updated_3d_marker_model()
+
+    def _run_optimization(self):
+        if not self.bg_task:
+            data_for_optimization = self.visibility_graphs.get_data_for_optimization()
+            if data_for_optimization:
+                args = (self.camera_model, data_for_optimization)
+                self.bg_task = background_helper.IPC_Logging_Task_Proxy(
+                    name="generator", generator=optimization_generator, args=args
+                )
+
+    def _get_updated_3d_marker_model(self):
```

Comment:
It is a bit weird that this function only returns something if the if-condition is True. It would be easier to read if there was an else-case that returns None explicitly.

Also this might be a nice place to utilize the `Observable` class: Currently we ask the optimizer in every iteration if there is a new model available. Most of the time the answer will be None, i.e. nothing happens. Instead of asking every time we could observe the optimizer and if a new model arrives execute a callback function that updates the model in the storage object for example. Let's discuss this when we integrate the Observable class with the UI!